core/src/main/kotlin/edu/stanford/spezi/core/DependenciesGraph.kt

```diff
@@ -39,8 +41,40 @@ class DependenciesGraph internal constructor(
      */
     inline fun <reified M : Module> dependency(identifier: String? = null): M {
         val typeKey = ModuleKey<M>(identifier)
-        return optionalDependency(typeKey)
-            ?: speziError("$typeKey not found. Please make sure to registered it the configuration block of your SpeziApplication")
+        return optionalDependency(typeKey) ?: createModuleOrThrow(typeKey)
+    }
+
+    /**
+     * Creates a module of type [M] if it was not previously registered in the dependency graph.
+     * It will attempt to instantiate the module by checking for an empty constructor, a constructor that takes a [Context] parameter, or
+     * finally check whether companion object of [M] implements [DefaultInitializer].
+     *
+     * @param typeKey The key associated with the instance built via [ConfigurationBuilder.module].
+     * @return The module instance of type [M] if it was previously registered or throws an error if not found.
+     */
+    @PublishedApi
+    @Suppress("UNCHECKED_CAST")
+    internal inline fun <reified M : Module> createModuleOrThrow(typeKey: ModuleKey<M>): M {
+        val clazz = M::class
+        val constructors = clazz.constructors
+        val appContext = optionalDependency<ApplicationModule>()?.application?.applicationContext
+        val result = runCatching {
+            constructors.find { it.parameters.isEmpty() }?.call()
+                ?: constructors.find { it.parameters.size == 1 && it.parameters[0].type.classifier == Context::class }?.call(appContext)
+                ?: run { appContext?.let { (clazz.companionObjectInstance as? DefaultInitializer<M>)?.create(it) } }
+                ?: speziError("No suitable constructor found for $typeKey")
+        }
+        val instance = result.getOrNull()
+        return if (instance != null) {
+            registry.modules[typeKey] = instance
+            logger.w { "Instantiated module $typeKey manually via fallback mechanism. Consider registering it explicitly." }
+            instance
+        } else {
+            speziError(
+                message = "$typeKey not found. Please make sure to register via in the configuration block of your app component",
+                cause = result.exceptionOrNull()
+            )
+        }
```

Comment:
What about constructors with default parameters, do they count as having less parameters, i.e. does the compiler just generate multiple versions of the same constructor with a separate set of parameters? 

I could for example imagine something like that, please let me know, if it doesn't make sense or I'm misinterpreting something:

```kotlin
class CredentialStorage(
    val context: Context, 
    val localStorage: LocalStorage = dependency<LocalStorage>()
) {}
```

This should still be automatically injected when needed, right? I've seen the approach with the create-function, but I'm just wondering, if we can create a more developer-friendly experience here.

Reply:
Okay - I'm fine with the annotation workaround in these cases, so all good - thank you!